_posts/2017-05-10-polyglot-scientific-computing-with-xtensor.md

```diff
@@ -0,0 +1,195 @@
+E logsumexp(const E& a, const axes_type& axes)
+{
+    auto m = xt::eval(xt::amax(a, axes));
+    auto sv = xt::slice_vector(m);
+    for (int i = 0; i < e1.dimension(); i++)
+    {
+        if (std::find (axes.begin(), axes.end(),  i) != axes.end())
+            sv.push_back(xt::newaxis());
+        else
+            sv.push_back(xt::all());
+    }
+    auto max2 = xt::dynamic_view(m, sv);
+    return E(m + xt::log(xt::sum(xt::exp(a - max2), axes)));
+}
+```
+
+With that change, we have two ways of exposing `logsumexp` to the Python. The first one is to directly instantiate the template:
+
+```cpp
+PYBIND11_PLUGIN(fast_lse)
+{
+    xt::import_numpy();
+    py::module m("fast_lse", "Fast implementation of logsumexp);
+    m.def("logsumexp", logsumexp<array_type>);
+    return m.ptr();
+}
+```
+
+However, if we decide to change the template parameter of `logsumexp` in the future, or if we add new template parameters, we'll have to update this binding code. To avoid this, we can wrap the call to the template `logsumexp` into a C++ function and let the template parameter deduction work for us:
+
+```cpp
+array_type python_logsumexp(const array_type& a, const axes_type& axes)
+{
+    return logsumexp(a, axes);
+}
+
+PYBIND11(fast_lse)
+{
+    xt::import_numpy();
+    py::module m("fast_lse", "Fast implementation of logsumexp);
+    m.def("logsumexp", python_logsumexp);
+    return m.ptr();
+}
+```
+
+Note that from Python, the function name remains unchanged even if the *implementation* name (i.e. `python_logsumexp`) has changed.
+
+## Enforcing the usage of expressions
+
+Our template `logsumexp` function is generic, however it allows to use any type for the first argument. If we try to call it with something that does not implement the `xexpression` interface, we will end up with cryptic errors, the compiler complaining about not being able to convert the parameter type to a very complicated template type.
+
+To avoid this, we can modify `logsumexp` signature so it complains earlier when its first parameter type is not an expression:
+
+```cpp
+template <class E>
+E logsumexp(const xt::xexpression<E>& e, const axes_type& axes)
+{
+    const E& a = e.derived_cast();
+    auto m = xt::eval(xt::amax(a, axes));
+    auto sv = xt::slice_vector(m);
+    for (int i = 0; i < e1.dimension(); i++)
+    {
+        if (std::find (axes.begin(), axes.end(),  i) != axes.end())
+            sv.push_back(xt::newaxis());
+        else
+            sv.push_back(xt::all());
+    }
+    auto max2 = xt::dynamic_view(m, sv);
+    return E(m + xt::log(xt::sum(xt::exp(a - max2), axes)));
+}
+```
+
+The first line in the function retrieves the real type of the expression, so that we can keep the rest of the implementation unchanged. Now you can see the benefits of wrapping `logsumexp` inside `python_logsumexp`: we don't have to change anything else in the code. Whereas if we had used direct instantiation of the template, the following changes would have been required:
```

Comment:
... in the code, while if we had used a ...

also, I find this sentence a bit unclear